File: java/src/LeetCode/LC0021_MergeTwoSortedLists.java
package LeetCode;

import MyStruct.ListNode;

public class LC0021_MergeTwoSortedLists {

    public static ListNode mergeTwoLists(ListNode list1, ListNode list2) {

        // 学会使用 dummy 头结点
        ListNode dummy = new ListNode(0);
        ListNode cur = dummy;

        while (list1 != null && list2 != null) {
            if (list1.val < list2.val) {
                cur.next = list1;
                cur = cur.next;
                list1 = list1.next;
            } else {
                cur.next = list2;
                cur = cur.next;
                list2 = list2.next;
            }
        }

        if (list1 != null) {
            cur.next = list1;
        }

        if (list2 != null) {
            cur.next = list2;
        }

        return dummy.next;
    }

}

File: cpp/Exercise/LC0021_MergeTwoSortedLists.cpp
#include "bits/stdc++.h"
#include "../MyStruct/ListNode.h"

using namespace std;

class Solution {
public:
    ListNode *mergeTwoLists(ListNode *list1, ListNode *list2) {

        ListNode *p = new ListNode(-1);
        ListNode *dummy = p;

        while (list1 != nullptr && list2 != nullptr) {
            if (list1->val < list2->val) {
                p->next = list1;
                list1 = list1->next;
            } else {
                p->next = list2;
                list2 = list2->next;
            }
            p = p->next;
        }

        if (list1 != nullptr) {
            p->next = list1;
        }

        if (list2 != nullptr) {
            p->next = list2;
        }

        return dummy->next;
    }
};